### PR TITLE
always fetch storage_ids by partition for latest materializations call

### PR DESCRIPTION
**Pull request by prha**
## Summary & Motivation
Switches the latest materializations by partition resolver to always use the optimized latest storage ids by partition call.

## How I Tested These Changes
BK

### CODE REVIEW

**Comment by prha** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/25511?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#25511** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/25511?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#25510** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/25510?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `master`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @prha and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: python_modules/dagster-graphql/dagster_graphql/schema/asset_graph.py [hunk 1]
```diff
@@ -980,42 +980,24 @@ def resolve_latestMaterializationByPartition(
...
-                ).values()
-            )
-            events_for_partitions = get_asset_materializations(
-                graphene_info,
-                asset_key=self._asset_node_snap.asset_key,
-                storage_ids=latest_storage_ids,
-            )
-            latest_materialization_by_partition = {
-                get_partition(event): event for event in events_for_partitions
-            }
```
Comment (gibsondan) [inline python_modules/dagster-graphql/dagster_graphql/schema/asset_graph.py:1004]: the change being made in this PR seems like a big improvement for the case where you pass in 1001 partitions and there are 10000 partitions - we were fetching a bunch of materializations before in that case that we then just threw out and didn't use right?

File: python_modules/dagster-graphql/dagster_graphql/schema/asset_graph.py [hunk 2]
```diff
@@ -980,42 +980,24 @@ def resolve_latestMaterializationByPartition(
...
-                    latest_materialization_by_partition[event_partition] = event
-                if len(latest_materialization_by_partition) == len(partitions):
-                    break
+        partitions = self.get_partition_keys() if partitions is None else partitions
+        latest_storage_ids = sorted(
+            (
+                graphene_info.context.instance.event_log_storage.get_latest_storage_id_by_partition(
+                    self._asset_node_snap.asset_key,
+                    DagsterEventType.ASSET_MATERIALIZATION,
+                    set(partitions),
```
Comment (gibsondan) [inline python_modules/dagster-graphql/dagster_graphql/schema/asset_graph.py:989]: do we need to pass this in the partitions set if partitions is None (i.e. we want all partitions)? I think we don't, right? I could imagine it being slower to add the filter that lists every single partition for that case

**Review by gibsondan — CHANGES_REQUESTED** [review]
one small thing - little worried about the 'fetch every asset partition' case and i think we could leave out passing in the filter for that case? otherwise looks good

**Comment by prha** [comment]
@gibsondan there's also an argument to be made that in the case partitions is `None`, we'll want to paginate through the known set of partitions in chunks of 1000...

File: python_modules/dagster-graphql/dagster_graphql/schema/asset_graph.py [hunk 3]
```diff
@@ -981,45 +981,27 @@ def resolve_latestMaterializationByPartition(
...
+                )
+            ).values()
+        )
+        events_for_partitions = get_asset_materializations(
+            graphene_info,
+            asset_key=self._asset_node_snap.asset_key,
+            storage_ids=latest_storage_ids,
+        )
+        latest_materialization_by_partition = {
+            get_partition(event): event for event in events_for_partitions
```
Comment (gibsondan) [inline python_modules/dagster-graphql/dagster_graphql/schema/asset_graph.py:999]: i don't think get_partition needs to be a lambda anymore now that it is only used in one path

**Review by gibsondan — APPROVED** [review]
re: paginating in chunks i could see that too, i guess we are trading off timing out the whole request potentially vs. timing out an individual query - could also see the argument for eventually disallowing the query_all_partitions case and requiring callsites to paginate, that might be better than paginating for them within the body of the resolver

**Pull request merged**